transports/quic/src/transport.rs

```diff
@@ -588,12 +613,34 @@ impl<P: Provider> Stream for Listener<P> {
                     let endpoint = self.endpoint.clone();
                     self.accept = async move { endpoint.accept().await }.boxed();
 
-                    let local_addr = socketaddr_to_multiaddr(&self.socket_addr(), self.version);
+                    let local_addr = socketaddr_to_multiaddr(
+                        &self.socket_addr(),
+                        self.version,
+                        self.web_transport_addr,
+                        Some(self.cert_hash),
+                    );
                     let remote_addr = connecting.remote_address();
-                    let send_back_addr = socketaddr_to_multiaddr(&remote_addr, self.version);
+                    //todo It's unclear should I add anything to a send_back_addr?
+                    let send_back_addr = socketaddr_to_multiaddr(
+                        &remote_addr,
+                        self.version,
+                        false,
+                        None,
+                    );
+
+                    let timeout = self.handshake_timeout.clone();
+                    let fut = if self.web_transport_addr {
```

Comment:
Instead of deciding this based on a boolean, it would be great if we could first accept the QUIC connection and dynamically decide based on what the client actually does.

In the current state, this means we cannot share a listener on the same port for WT and regular QUIC connections.

Reply:
If I understand the `quinn` docs correctly, we can do `connecting.await` which gives us a `quinn::Connection`. We can then call [`Connection::handshake_data`](https://docs.rs/quinn/latest/quinn/struct.Connection.html#method.handshake_data) and downcast that to https://docs.rs/quinn/latest/quinn/crypto/rustls/struct.HandshakeData.html.

From there, we can inspect the ALPN protocol and decide whether this is a regular libp2p QUIC connection or in fact a WT transport connection.

Using that, we can then branch correctly on what we want to do, assuming the user is actually listening on both transports! If the user only said `listen_on` without `/webtransport`, then we should reject the connection. Similarly, if we are only listening on `/webtransport`, we should (probably?) reject regular QUIC connections.

I think to achieve this, we will need to slightly modify our `listen_on` behaviour to remember, which port we are listening on and try to find an existing listener on the same port to _add_ the specific functionality.

Reply:
> In the current state, this means we cannot share a listener on the same port for WT and regular QUIC connections.

It would be really nice if we could achieve this but I'd also be okay if we land this in a 2nd iteration.